### PR TITLE
move to predict_on_batch

### PR DESCRIPTION
**Pull request by jeromelecoq**
The goal of this PR is to move to using predict_on_batch for inference since that should be both memory and computation efficient.

### CODE REVIEW

**Comment by jeromelecoq** [comment]
@danielsf @aamster any concerns with this simple PR?

**Comment by aamster** [comment]
@jeromelecoq I am not sure I understand why this fixes the issue. `local_data[0]` is a minibatch of data of type `Sequence`, correct? So `predict` will iterate through the dataset, but `predict_on_batch` will just predict on the entire minibatch without iterating. It seems like `predict` might be slower if the input is larger, but shouldn't use more memory.

The keras docs say to call the model directly for small inputs, ie `self.model(local_data[0], training=False)`. Does that also fix the issue?

**Comment by jeromelecoq** [comment]
The issue is I think due to a memory leak. I found that looping with predict and GPUs will try to create huge arrays on the GPU. see here: https://github.com/AllenInstitute/deepinterpolation/issues/77

**Comment by jeromelecoq** [comment]
It also seemed like this was the expected use in our case, since I am actually feeding one batch of data as described here : https://stackoverflow.com/questions/44972565/what-is-the-difference-between-the-predict-and-predict-on-batch-methods-of-a-ker

**Comment by aamster** [comment]
@jeromelecoq I am still confused as to why `predict` is running into a memory issue if `local_data[0]` is only a single minibatch, since, as the docs say, the default batch_size on `predict` is 32, and I believe `local_data[0]` has a batch size of 5, correct? In which case it should just take the entire input and predict on it, which is the same as `predict_on_batch`. 

Are you able to try using `self.model(local_data[0], training=False)` which the docs suggest for small inputs? If not then it is fine.

**Comment by jeromelecoq** [comment]
> @jeromelecoq I am still confused as to why `predict` is running into a memory issue if `local_data[0]` is only a single minibatch, since, as the docs say, the default batch_size on `predict` is 32, and I believe `local_data[0]` has a batch size of 5, correct? In which case it should just take the entire input and do a forward pass on it, which is the same as `predict_on_batch`.

Yes batch size is usually set to 5 but in this case it is set by the generator so is user-defined. I am with you on not knowing why predict gives a GPU memory issue given the doc. But I can read from the doc that looping predict 1000 times might not be what they are testing for since predict is designed to break down batches internally. In addition, the array allocation request in the error is completely unexpected given the network architecture. 

> Are you able to try using `self.model(local_data[0], training=False)` which the docs suggest for small inputs? If not then it is fine.

I can try that, why do you think that would be superior to .preedict_on_batch? Can you point me at the doc where this is described? 
Thanks!

**Comment by aamster** [comment]
[Here](https://www.tensorflow.org/api_docs/python/tf/keras/Model#predict) it says

it says
> For small amount of inputs that fit in one batch, directly using __call__() is recommended for faster execution, e.g., model(x), or model(x, training=False) if you have layers such as tf.keras.layers.BatchNormalization that behaves differently during inference.

I am just worried that they don't mention `predict_on_batch` in their docs here, though it is probably fine.

**Comment by jeromelecoq** [comment]
Ah I see, I am also confused now why do they have predict_on_batch and call options?
